plugins/color/gsd-color-state.c

```diff
@@ -46,6 +47,70 @@
 
 static void gcm_session_set_gamma_for_all_devices (GsdColorState *state);
 
+static guint gcm_session_get_output_percentage (void);
+
+typedef struct {
+  unsigned short vendor;
+  unsigned short product;
+} EdidInfo;
+
+typedef struct {
+  unsigned long pattern;
+  EdidInfo info;
+} EdidHeader;
+
+static EdidInfo
+known_oled[] = {
+  // SDC OLED panel in S76 Adder and XPS 7590
+  { 19587, 41001 }
+};
+
+static gboolean
+has_oled_builtin_display (void)
+{
+  gboolean has_oled = FALSE;
+
+  // TODO: The Rust code does a dir walk to find all outputs. Needed?
+  FILE *f = fopen("/sys/class/drm/card0-eDP-1/edid", "rb");
+  if (f == NULL) {
+    return FALSE;
+  }
+
+  // Only need the first 12 bytes of the header
+  char *edid = malloc(12);
+  size_t read_bytes = fread(edid, 1, 12, f);
+  fclose(f);
+
+  if (read_bytes == 12) {
+    EdidHeader header = { };
+    memcpy(&header.pattern, edid, 8);
+    memcpy(&header.info.vendor, edid + 8, 2);
+    memcpy(&header.info.product, edid + 10, 2);
+
+    // TODO: Endianess swap. Swap here or change the value in the array?
+    unsigned short vendor = header.info.vendor;
+    header.info.vendor = (vendor >> 8) | (vendor << 8);
+    
+    if (header.pattern == 0xFFFFFFFFFFFF00) {
+      // TODO: I don't like this line. Alternatives?
+      unsigned long known_oled_len = sizeof(known_oled) / sizeof(known_oled[0]);
```

Comment:
This is a pretty standard way to get the size of a statically defined array in C, you can remove the TODO

Reply:
Thanks! I still have a hard time distinguishing kludges from standard C.